utils/s3/client.cc

```diff
@@ -1039,6 +1039,141 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::download_source final : public seastar::data_source_impl {
+    static constexpr size_t default_part_size = 10_MiB;
+
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+
+    using part_buffers = circular_buffer<temporary_buffer<char>>;
```

Comment:
I don't understand why we need this buffer. Isn't the tcp read queue already acting as an auto-sizing read-ahead buffer?

I think we can just read from the stream and produce buffers.

http client -> temporary buffer -> data_source -> input_stream

Reply:
We can't just return http::content_length_source_impl to the client, in case S3 server closes its socket in the middle, the GET should be replayed with reduced Range header, so an interposing data_source is needed

Reply:
Makes sense. But do we need any buffering apart from that provided by upstream data_source:s and tcp?

Reply:
Good question. Maybe we don't indeed